libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -234,12 +234,73 @@ void IRGeneratorForStatements::endVisit(Return const& _return)
 
 void IRGeneratorForStatements::endVisit(UnaryOperation const& _unaryOperation)
 {
-	if (type(_unaryOperation).category() == Type::Category::RationalNumber)
+	Type const& resultType = type(_unaryOperation);
+	Token const op = _unaryOperation.getOperator();
+
+	if (resultType.category() == Type::Category::RationalNumber)
+	{
 		defineExpression(_unaryOperation) <<
-			formatNumber(type(_unaryOperation).literalValue(nullptr)) <<
+			formatNumber(resultType.literalValue(nullptr)) <<
 			"\n";
+	}
+	else if (resultType.category() == Type::Category::Integer)
+	{
+		string modifiedValue = m_context.newYulVariable();
+
+		if (op == Token::Inc || op == Token::Dec)
+		{
+			IntegerType const& intType = *dynamic_cast<IntegerType const*>(&resultType);
+
+			solAssert(!!m_currentLValue, "LValue not retrieved.");
+			string fetchValueExpr = m_currentLValue->retrieveValue();
+			string modifiedValue = m_context.newYulVariable();
+			string originalValue = m_context.newYulVariable();
+
+			m_code << Whiskers(R"(
+				let <originalValue> := <fetchExpr>
+				let <modifiedValue> := <incrOrdecr>(<originalValue>)
+				<storeValue>
+			)")
+				("originalValue", originalValue)
+				("fetchExpr", fetchValueExpr)
+				("modifiedValue", modifiedValue)
+				("incrOrdecr",op == Token::Inc ?
+					m_utils.incrementCheckedFunction(intType) :
+					m_utils.decrementCheckedFunction(intType))
+				("storeValue", m_currentLValue->storeValue(modifiedValue, resultType))
+				.render();
+
+			defineExpression(_unaryOperation) <<
+				(_unaryOperation.isPrefixOperation() ? modifiedValue : originalValue) << "\n";
+
+			m_code << m_currentLValue->storeValue(modifiedValue, resultType);
+			m_currentLValue.reset();
+		}
+		else if (op == Token::BitNot)
+			appendSimpleUnaryOperation(_unaryOperation, modifiedValue, _unaryOperation.subExpression());
```

Comment:
I'm not sure it is that simple, because it will set higher-order bits that have to be reset to zero after the operation to stay inside the range of the type. Could use some existing cleanup function, maybe.